constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Windows.java

```diff
@@ -1,267 +1,334 @@
 package gov.nasa.jpl.aerie.constraints.time;
 
 import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import org.apache.commons.lang3.tuple.Pair;
 
-import java.util.Iterator;
+import java.util.Arrays;
 import java.util.List;
-import java.util.Objects;
 import java.util.Optional;
-import java.util.stream.Collectors;
+import java.util.Spliterator;
+import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
-import static gov.nasa.jpl.aerie.constraints.time.Window.Inclusivity.Inclusive;
+public final class Windows extends IntervalMap<Boolean> {
```

Comment:
`set` and `unset` (and their variants), as well as `bound` are used by clients. `map` and `map2` are not. `min/maxValidTimePoint` aren't used yet by clients but they very well could be. 

Its in the definition: Windows *is* (not has) a boolean profile, with some special operations. It does not have any data specific to the `Windows` class. It is basically nothing more than an extension trait just for `IntervalMap<Boolean>`; and the Java version of an extension trait is a subclass with no extra fields. If we eventually convert other profiles to use IntervalMap, the same would be true of them. A string profile would be `IntervalMap<String>` with an added `concatenate` operation. A linear profile would be `IntervalMap<LinearProfilePiece>` with added comparison and arithmetic operations.

Reply:
I've rarely found "is-a / has-a" reasoning to have much bearing on the maintenance and usability aspects of a design, unfortunately. The fact remains that immutable methods conflicts with subclassing, since the superclass methods returns values of the superclass, not of the subclass.

My concerns about mutation are primary, I think; we should probably discuss that before continuing on the inheritance point.

Reply:
> "is-a / has-a" reasoning

Just to riff on this point a little more, I would say that windows are *defined by* an interval map of booleans; the "is-a" and "has-a" distinction only enters the picture as primarily an implementation detail. We need to provide an appropriate interface for clients of the class either way. Mechanically, inheritance lets us reuse code and state in a particular way, but it's not the only way, and I tend to find its way fraught more often than not.

Reply:
No, it has significant impact on the maintenance. If we ever consider using IntervalMap for more than booleans, then every method in IntervalMap (except `equals` and `toString`), including `map`, will need to be delegated. It will lead to more coupled code, not less, because `Windows` and every other specialization of `IntervalMap` will need duplicate delegation to all of the methods of IntervalMap.

Reply:
If a client uses a method that returns a new `IntervalMap`, the client leaves the zone of specialization already. This effectively makes immutable methods terminal, unless you then take the returned map and feed it to a `Windows` constructor to re-wrap it, or `@Override` the immutable methods to wrap `super` calls (which ruins the automatic delegation we get from inheritance). The subclassing approach only gains if you commit fully to a mutation-based API, or use immutable methods solely as terminal operations for which you're okay with a generic `IntervalMap` result. This is why I think we should sort the mutation question first.

Alternatively, we can commit to using `IntervalMap<V>` everywhere, and making the methods on `Windows` and other specializations static helper methods. This lets us keep a mix of mutating and non-mutating methods on `IntervalMap` (since we're consistently using the same type), while also letting us define capabilities for particular instantiations of the `IntervalMap` type parameter.

Reply:
resolving; decision was to use `IntervalMap` directly and have `Windows` provide static methods.